File: android/src/main/java/cn/nodemedia/react_native_nodemediaclient/RCTNodeCameraView.java
//
//  RCTNodeCameraView.java
//
//  Created by Mingliang Chen on 2017/11/29.
//  Copyright © 2017年 NodeMedia. All rights reserved.
//

package cn.nodemedia.react_native_nodemediaclient;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.util.Base64;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.RCTEventEmitter;

import java.io.ByteArrayOutputStream;

import cn.nodemedia.NodeCameraView;
import cn.nodemedia.NodePublisher;
import cn.nodemedia.NodePublisherDelegate;

public class RCTNodeCameraView extends NodeCameraView implements LifecycleEventListener {
    private NodePublisher mNodePublisher;
    private Boolean isAutoPreview = false;

    private int cameraId = -1;
    private boolean cameraFrontMirror = true;

    private int audioBitrate = 32000;
    private int audioProfile = 0;
    private int audioSamplerate = 44100;
    private static String CURRENT_FRAME_EVENT_NAME = "CURRENT_FRAME_EVENT_NAME";

    private int videoPreset = NodePublisher.VIDEO_PPRESET_4X3_480;
    private int videoFPS = 20;
    private int videoBitrate = 400000;
    private int videoProfile = NodePublisher.VIDEO_PROFILE_BASELINE;
    private boolean videoFrontMirror = false;

    private boolean denoise = false;
    private int smoothSkinLevel = 0;


    public RCTNodeCameraView(@NonNull ThemedReactContext context) {
        super(context);
        context.addLifecycleEventListener(this);

        mNodePublisher = new NodePublisher(context, RCTNodeMediaClient.getPremium());
        mNodePublisher.setNodePublisherDelegate(new NodePublisherDelegate() {
            @Override
            public void onEventCallback(NodePublisher nodePublisher, int i, String s) {
                WritableMap event = Arguments.createMap();
                event.putInt("code", i);
                event.putString("message", "s");
                ReactContext reactContext = (ReactContext) getContext();
                reactContext.getJSModule(RCTEventEmitter.class).receiveEvent(
                        getId(),
                        "topChange",
                        event);
            }
        });

    }

    public void setOutputUrl(String url) {
        mNodePublisher.setOutputUrl(url);

    }


    public void setCamera(int cameraId, boolean cameraFrontMirror) {
        mNodePublisher.setCameraPreview(this, cameraId, cameraFrontMirror);
        if(isAutoPreview) {
            startPrev();
        }
    }

    public void setAudio(int audioBitrate, int audioProfile,int audioSamplerate) {
        mNodePublisher.setAudioParam(audioBitrate, audioProfile, audioSamplerate);
    }

    public void setVideo(int videoPreset, int videoFPS, int videoBitrate, int videoProfile, boolean videoFrontMirror ) {
        mNodePublisher.setVideoParam(videoPreset, videoFPS, videoBitrate, videoProfile, videoFrontMirror);
    }

    public void setDenoise(boolean denoise) {
        this.denoise = denoise;
    }

    public void setSmoothSkinLevel(int smoothSkinLevel) {
        this.smoothSkinLevel = smoothSkinLevel;
    }

    public int startPrev() {
        return mNodePublisher.startPreview();
    }

    public int stopPrev() {
        return mNodePublisher.stopPreview();
    }

    public int start() {
        return mNodePublisher.start();
    }

    public void stop() {
        mNodePublisher.stop();
    }

    public int switchCam() {
        return mNodePublisher.switchCamera();
    }

    public void captureCurrentViewAsBase64(final int quality){

        mNodePublisher.capturePicture(new NodePublisher.CapturePictureListener() {
            @Override
            public void onCaptureCallback(Bitmap bitmap) {
                ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
                bitmap.compress(Bitmap.CompressFormat.JPEG, quality, byteArrayOutputStream);
                byte[] byteArray = byteArrayOutputStream .toByteArray();
                String encoded = Base64.encodeToString(byteArray, Base64.DEFAULT);
                WritableMap params = Arguments.createMap(); // add here the data you want to send
                params.putString("base64",encoded);
                ((ReactContext) getContext())
                        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                        .emit(CURRENT_FRAME_EVENT_NAME,params);
            }
        });



    }
    public void audioPreview() {
        isAutoPreview = true;
        if(cameraId >=0) {
            startPrev();
        }

    }

    @Override
    public void onHostResume() {

    }

    @Override
    public void onHostPause() {

    }

    @Override
    public void onHostDestroy() {
        mNodePublisher.stopPreview();
        mNodePublisher.stop();
    }
}

File: ios/RCTNodeMediaClient/RCTNodeCameraView.m
//
//  RCTNodeCameraView.m
//  
//
//  Created by Mingliang Chen on 2017/12/12.
//  Copyright © 2017年 NodeMedia. All rights reserved.
//

#import "RCTNodeMediaClient.h"
#import "RCTNodeCameraView.h"
#import <React/RCTUIManager.h>
#import <AVFoundation/AVFoundation.h>

@interface RCTNodeCameraView()

@property (strong,nonatomic) NodePublisher *np;

@end

@implementation RCTNodeCameraView

-(id)initWithBridge:(RCTBridge *)bridge{
    self = [super init];
    if(self) {
        _np = [[NodePublisher alloc] initWithPremium:[RCTNodeMediaClient premium]];
        [_np setNodePublisherDelegate:self];
        _autopreview = NO;
        _outputUrl = nil;
        _camera = nil;
        _audio = nil;
        _video = nil;
    }
    [self setBridge:bridge];
    return self;
}

-(void)onEventCallback:(id)sender event:(int)event msg:(NSString *)msg{
    NSDictionary *data = @{@"code": [NSNumber numberWithInt:event],  @"message": msg,
                           @"target": self.reactTag};
    [self.bridge.eventDispatcher sendInputEventWithName:@"topChange" body:data];
}

-(void)setOutputUrl:(NSString *)outputUrl {
  [_np setOutputUrl:outputUrl];
}

-(void)setAutopreview:(BOOL)autopreview {
  _autopreview = autopreview;
  if(_camera && _video && autopreview) {
    [_np startPreview];
  }
}
- (void)setCamera:(NSDictionary *)camera {
  _camera = camera;
  int cameraId = [[camera objectForKey:@"cameraId"] intValue];
  BOOL cameraFrontMirror = [[camera objectForKey:@"cameraFrontMirror"] boolValue];
  [_np setCameraPreview:self cameraId:cameraId frontMirror:cameraFrontMirror];
  if(_autopreview && _video) {
    [_np startPreview];
  }
}

- (void)setAudio:(NSDictionary *)audio {
  _audio = audio;
  int audioBitrate = [[audio objectForKey:@"bitrate"] intValue];
  int audioProfile = [[audio objectForKey:@"profile"] intValue];
  int audioSamplerate = [[audio objectForKey:@"samplerate"] intValue];
   
  if (audioSamplerate == 0){
      AVAudioSession* session = [AVAudioSession sharedInstance];
      BOOL success;
      NSError* error = nil;
      double preferredSampleRate = 48000;
      success  = [session setPreferredSampleRate:preferredSampleRate error:&error];
      if (success) {
          audioSamplerate = preferredSampleRate;
      } else {
          NSLog (@"error setting sample rate %@", error);
      }
  }
  [_np setAudioParamBitrate:audioBitrate profile:audioProfile sampleRate:audioSamplerate];
}

- (void)setVideo:(NSDictionary *)video {
  _video = video;
  int videoPreset = [[video objectForKey:@"preset"] intValue];
  int videoFPS = [[video objectForKey:@"fps"] intValue];
  int videoBitrate = [[video objectForKey:@"bitrate"] intValue];
  int videoProfile = [[video objectForKey:@"profile"] intValue];
  BOOL videoFrontMirror = [[video objectForKey:@"videoFrontMirror"] boolValue];
  [_np setVideoParamPreset:videoPreset fps:videoFPS bitrate:videoBitrate profile:videoProfile frontMirror:videoFrontMirror];
  if(_autopreview && _camera) {
    [_np startPreview];
  }
}

- (void)setDenoise:(BOOL)denoise {
  _denoise = denoise;
  [_np setDenoiseEnable:denoise];
}

- (void)setSmoothSkinLevel:(NSInteger)smoothSkinLevel {
  _smoothSkinLevel = smoothSkinLevel;
  [_np setBeautyLevel:smoothSkinLevel];
}

- (void)setFlashEnable:(BOOL)flashEnable {
  [_np setFlashEnable:flashEnable];
}

-(void)captureCurrentFrame:(float)quality{
    [_np capturePicture:^(UIImage * _Nullable image) {
        if(image){
            NSString * base64 = [UIImageJPEGRepresentation(image,quality) base64EncodedStringWithOptions:NSDataBase64Encoding64CharacterLineLength];
            [self.bridge.eventDispatcher sendDeviceEventWithName:@"currentFrameUpdate" body:@{@"base64": base64}];
        }
        
     }
     ];
}
-(int)startprev {
  return [_np startPreview];
}

-(int)stopprev {
  return [_np stopPreview];
}

-(int)start {
    int res = 0;
    @try{
      res =[_np start];
    }
    @catch(NSException * ex){
        NSLog(@"%@",ex.reason);
    }
    return res;
}

-(int)stop {
  return [_np stop];
}

-(int)switchCamera {
  return [_np switchCamera];
}

@end
